File: tl/src/main/java/com/github/badoualy/telegram/tl/api/TLPaymentRequestedInfo.java
package com.github.badoualy.telegram.tl.api;

import com.github.badoualy.telegram.tl.TLContext;
import com.github.badoualy.telegram.tl.core.TLObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLObject;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLObject;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32;
import static com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLPaymentRequestedInfo extends TLObject {

    public static final int CONSTRUCTOR_ID = 0x909c3f94;

    protected int flags;

    protected String name;

    protected String phone;

    protected String email;

    protected TLPostAddress shippingAddress;

    private final String _constructor = "paymentRequestedInfo#909c3f94";

    public TLPaymentRequestedInfo() {
    }

    public TLPaymentRequestedInfo(String name, String phone, String email, TLPostAddress shippingAddress) {
        this.name = name;
        this.phone = phone;
        this.email = email;
        this.shippingAddress = shippingAddress;
    }

    private void computeFlags() {
        flags = 0;
        flags = name != null ? (flags | 1) : (flags & ~1);
        flags = phone != null ? (flags | 2) : (flags & ~2);
        flags = email != null ? (flags | 4) : (flags & ~4);
        flags = shippingAddress != null ? (flags | 8) : (flags & ~8);
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        computeFlags();

        writeInt(flags, stream);
        if ((flags & 1) != 0) {
            if (name == null) throwNullFieldException("name", flags);
            writeString(name, stream);
        }
        if ((flags & 2) != 0) {
            if (phone == null) throwNullFieldException("phone", flags);
            writeString(phone, stream);
        }
        if ((flags & 4) != 0) {
            if (email == null) throwNullFieldException("email", flags);
            writeString(email, stream);
        }
        if ((flags & 8) != 0) {
            if (shippingAddress == null) throwNullFieldException("shippingAddress", flags);
            writeTLObject(shippingAddress, stream);
        }
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        flags = readInt(stream);
        name = (flags & 1) != 0 ? readTLString(stream) : null;
        phone = (flags & 2) != 0 ? readTLString(stream) : null;
        email = (flags & 4) != 0 ? readTLString(stream) : null;
        shippingAddress = (flags & 8) != 0 ? readTLObject(stream, context, TLPostAddress.class,
                                                          TLPostAddress.CONSTRUCTOR_ID) : null;
    }

    @Override
    public int computeSerializedSize() {
        computeFlags();

        int size = SIZE_CONSTRUCTOR_ID;
        size += SIZE_INT32;
        if ((flags & 1) != 0) {
            if (name == null) throwNullFieldException("name", flags);
            size += computeTLStringSerializedSize(name);
        }
        if ((flags & 2) != 0) {
            if (phone == null) throwNullFieldException("phone", flags);
            size += computeTLStringSerializedSize(phone);
        }
        if ((flags & 4) != 0) {
            if (email == null) throwNullFieldException("email", flags);
            size += computeTLStringSerializedSize(email);
        }
        if ((flags & 8) != 0) {
            if (shippingAddress == null) throwNullFieldException("shippingAddress", flags);
            size += shippingAddress.computeSerializedSize();
        }
        return size;
    }

    @Override
    public String toString() {
        return _constructor;
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public TLPostAddress getShippingAddress() {
        return shippingAddress;
    }

    public void setShippingAddress(TLPostAddress shippingAddress) {
        this.shippingAddress = shippingAddress;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/TLPaymentRequestedInfo.kt
package com.github.badoualy.telegram.tl.api

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_BOOLEAN
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.core.TLObject
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException
import kotlin.Any
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.jvm.Throws

/**
 * paymentRequestedInfo#909c3f94
 *
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLPaymentRequestedInfo() : TLObject() {
    var name: String? = null

    var phone: String? = null

    var email: String? = null

    var shippingAddress: TLPostAddress? = null

    private val _constructor: String = "paymentRequestedInfo#909c3f94"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            name: String?,
            phone: String?,
            email: String?,
            shippingAddress: TLPostAddress?
    ) : this() {
        this.name = name
        this.phone = phone
        this.email = email
        this.shippingAddress = shippingAddress
    }

    protected override fun computeFlags() {
        _flags = 0
        updateFlags(name, 1)
        updateFlags(phone, 2)
        updateFlags(email, 4)
        updateFlags(shippingAddress, 8)
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        computeFlags()

        writeInt(_flags)
        doIfMask(name, 1) { writeString(it) }
        doIfMask(phone, 2) { writeString(it) }
        doIfMask(email, 4) { writeString(it) }
        doIfMask(shippingAddress, 8) { writeTLObject(it) }
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        _flags = readInt()
        name = readIfMask(1) { readString() }
        phone = readIfMask(2) { readString() }
        email = readIfMask(4) { readString() }
        shippingAddress = readIfMask(8) { readTLObject<TLPostAddress>(TLPostAddress::class, TLPostAddress.CONSTRUCTOR_ID) }
    }

    override fun computeSerializedSize(): Int {
        computeFlags()

        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += getIntIfMask(name, 1) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(phone, 2) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(email, 4) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(shippingAddress, 8) { it.computeSerializedSize() }
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLPaymentRequestedInfo) return false
        if (other === this) return true

        return _flags == other._flags
                && name == other.name
                && phone == other.phone
                && email == other.email
                && shippingAddress == other.shippingAddress
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0x909c3f94.toInt()
    }
}
